[feat & fix] 은퇴 목표 default 생성 기능 구현 & 내 목표 전체 조회 기능 고도화

rlqor9688 opened the pull request:
## 📌 관련 이슈
- Closes #69 
- Closes #70 

## ✨ 변경 사항
… / Redis timeout, retry 설정 추가

[Build.gradle]
1. Spring retry 의존성 추가

[Application]
1. @EnableRetry annotation 추가

[Config]
1. RestTemplateConfig : ConnectTimeout, ReadTimeout 세팅 추가

[Controller]
1. getMyGoals 메서드 내 RequestParam(Cusror 및 필터 조건) 추가

[Dto]
1. GoalRetirementCalculateRequest : Positive 조건 추가
2. GoalRetirementEstimateResponse : expectedRetirementDate 반환 필드 추가
3. GoalRetirementRequest : 기대 수명 Positive 조건 추가
4. CursorPageResponse 클래스 설정
5. GoalSearchCondition : 조회 조건 record 설정
6. GoalUpdateRequest : startAt NotNull 조건 추가 / @ValidGoalPeriod Annotation 추가

[Service]
1. GoalRetirementCalculateService : 예외 조건 방어 로직 추가
2. GoalService : findMyGoals 반환 타입 변경(커서)
3. GoalServiceImpl : 은퇴 목표가 하나 있을 때 생성하지 못하도록 하는 방어 로직 구현, findMyGoals 커서 및 필터 조회 기능으로 수정
4. OpenApiApartmentPriceService : Retry annotation 및 설정 추가
5. RetirementReferenceValueService : Response 내 예상 은퇴 시점 필드 추가
7. DefaultRetirementGoalService : 은퇴 목표 생성 로직 구현
8. DefaultRetirementGoalServiceTest : 은퇴목표 생성 로직 테스트코드 작성

[Test]
1. GoalServiceImplTest : findMyGoals 수정 로직 검증 테스트 코드 구현
2. GoalSummaryResponse 내 id 추가

[Entity]
1. Goal : 정적팩토리메서드 내 status, commentCount, likeCount 필드 추가

[Repository]
1. GoalRepository
- CustomGoalRepository 확장성 추가
- existsByUserIdAndCategory 메서드 추가
- countByUserIdAndCategory 메서드 추가
2. RedisApartmentPriceRepository : 주석 수정
3. CustomGoalRepository : searchGoalsWithCursor 메서드 정의
4. CustomGoalRepositoryImpl : searchGoalsWithCursor 메서드 구현

## 📷 Postman 
- GoalServiceTestImplTest
![image](https://github.com/user-attachments/assets/24d87851-33d6-40b2-875f-0b982bba4c1c)

- DefaultRetirementGoalServiceTest
![image](https://github.com/user-attachments/assets/cfdf622a-fe84-43ce-96b0-679dd3fa407f)

## ✅ 체크리스트
- [ ] 관련 이슈에 연결됨
- [ ] 기능이 정상적으로 동작함
- [ ] 코드 리뷰를 받았음
- [ ] 불필요한 로그/코드 제거됨

## 💬 기타 참고 사항
- 테스트 방법, 추가 설명 등 필요한 경우 작성해 주세요.

rlqor9688 committed:
[feat & fix] 은퇴 목표 default 생성 기능 구현 & 내 목표 전체 조회 Cursor, filter 기능 구현 / Redis timeout, retry 설정 추가
[Build.gradle]
1. Spring retry 의존성 추가

[Application]
1. @EnableRetry annotation 추가

[Config]
1. RestTemplateConfig : ConnectTimeout, ReadTimeout 세팅 추가

[Controller]
1. getMyGoals 메서드 내 RequestParam(Cusror 및 필터 조건) 추가

[Dto]
1. GoalRetirementCalculateRequest : Positive 조건 추가
2. GoalRetirementEstimateResponse : expectedRetirementDate 반환 필드 추가
3. GoalRetirementRequest : 기대 수명 Positive 조건 추가
4. CursorPageResponse 클래스 설정
5. GoalSearchCondition : 조회 조건 record 설정

[Service]
1. GoalRetirementCalculateService : 예외 조건 방어 로직 추가
2. GoalService : findMyGoals 반환 타입 변경(커서)
3. GoalServiceImpl : 은퇴 목표가 하나 있을 때 생성하지 못하도록 하는 방어 로직 구현, findMyGoals 커서 및 필터 조회 기능으로 수정
4. OpenApiApartmentPriceService : Retry annotation 및 설정 추가
5. RetirementReferenceValueService : Response 내 예상 은퇴 시점 필드 추가
6. DefaultRetirementGoalService : 은퇴 목표 생성 로직 구현
7. DefaultRetirementGoalServiceTest : 은퇴목표 생성 로직 테스트코드 작성

[Test]
1. GoalServiceImplTest : findMyGoals 수정 로직 검증 테스트 코드 구현
2. GoalSummaryResponse 내 id 추가

[Entity]
1. Goal : 정적팩토리메서드 내 status, commentCount, likeCount 필드 추가

[Repository]
1. GoalRepository
- CustomGoalRepository 확장성 추가
- existsByUserIdAndCategory 메서드 추가
- countByUserIdAndCategory 메서드 추가
2. RedisApartmentPriceRepository : 주석 수정
3. CustomGoalRepository : searchGoalsWithCursor 메서드 정의
4. CustomGoalRepositoryImpl : searchGoalsWithCursor 메서드 구현

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-goal

rlqor9688 committed:
[docs] GoalUpdateRequest 내 @ValidGoalPeriod 검증 추가
[Dto]
1. GoalUpdateRequest : NotNull 검증 조건 추가

ChangHoonS commented on src/main/java/org/example/lifechart/common/config/RestTemplateConfig.java:
현재 factory.setConnectTimeout(3000);과 factory.setReadTimeout(5000);를 넣으셨는데
RestTemplate에 만든 factory를 넣어주지 않으면 타임아웃 반영이 안된다고 하는 것 같습니다.
확인 부탁 드립니다.

rlqor9688 replied:
네 확인했습니다! 감사합니다! 반영해서 올리겠습니다!

rlqor9688 replied:
이 부분 Factory 의존성 주입에서 충돌이 생겨, timeout 배제하고 올린 뒤 고도화 작업으로 업데이트 하겠습니다.

ChangHoonS commented on src/main/java/org/example/lifechart/domain/goal/controller/GoalController.java:
@RequestParam(required = false, defaultValue = "10") Integer size
여기서 디폴트 값이 있으면 true가 되어도 상관없는 것 아닌가요?
아니면 false에 그냥 따로 디폴트 값을 넣은건가요? 궁금합니다.

rlqor9688 replied:
클라이언트가 size를 보내주지 않아도 default로 10을 넣어주겠다는 의미입니다!

ChangHoonS replied:
아 제가 잠시 헷갈렸네요

ChangHoonS commented on src/main/java/org/example/lifechart/domain/goal/dto/request/GoalRetirementRequest.java:
@Positive 넣으셨는데 여기는 따로 에러 뜨면 메세지는 넣지 않아도 될까요?

rlqor9688 replied:
오 맞네요! 감사합니다! 반영하겠습니다!

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-goal

rlqor9688 committed:
[fix] 코드 컨벤션 통일 및 Redis/Retry 설정 추가
[Enum]
1. @Getter, @RequiredArgsConstructor 구조로 통일

[Config]
1. RestTemplateConfig : HttpComponentsClientHttpRequestFactory 설정 이슈로 임시 삭제
2. RetryConfig : Api연결 재시도를 위한 Config 설정

[Dto]
1. GoalRetirementCalculateRequest, GoalRetirementRequest 내 @Positive 메시지 추가

[Build]
1. Retry 설정 추가

rlqor9688 committed:
[fix] spring.datasource.url 오타 수정

woneyee commented on src/main/java/org/example/lifechart/domain/goal/service/DefaultRetirementGoalService.java:
```diff
@@ -0,0 +1,70 @@
+
+		LocalDateTime startAt = LocalDateTime.now();
+
+		GoalRetirementEstimateResponse retirementResponse = retirementReferenceValueService.getReferenceValues(userId, startAt.getYear());
```
이 부분 GoalRetirementCalculateRequest calculateRequest = GoalRetirementCalculateRequest.builder()
	.startAt(startAt).endAt(retirementResponse.getExpectedRetirementDate()) 이런 식으로 고치면 코드가 더 간결해질 것 같습니다!!

rlqor9688 replied:
피드백 감사합니다!
```
LocalDateTime endAt = retirementResponse.getExpectedRetirementDate();
		Long expectedLifespan = retirementResponse.getExpectedLifespan();
		Long monthlyExpense = retirementResponse.getMonthlyExpense();
		RetirementType retirementType = retirementResponse.getRetirementType();
```
이 부분을 이렇게 인스턴스로 정의하지 않고, 아래의 빌더 패턴의 입력 필드안에 바로 넣어주면 간결해질 것 같다는 말씀이 맞을까요~?

woneyee replied:
맞습니답

rlqor9688 replied:
피드백 감사합니다! 반영했습니다!

woneyee commented on src/main/java/org/example/lifechart/domain/goal/repository/GoalRepository.java:
```diff
@@ -23,6 +27,10 @@ public interface GoalRepository extends JpaRepository<Goal, Long>, CustomShareGo
+
+    Long countByUserIdAndCategory(Long userId, Category category);
+
     @Query("""
```
쿼리가 중간에 끊긴건가요??

rlqor9688 replied:
@Query("""SELECT g FROM Goal g
    JOIN FETCH g.user
    WHERE g.id IN :goalIds AND g.user.id = :userId
    """)
    List<Goal> findAllWithUserByIdAndUserId(@Param("goalIds") List<Long> goalIds, @Param("userId") Long userId);
    
    이렇게 있는 것 같습니다!

woneyee reviewed (approved):
넵 저는 확인끝났습니다 제가 너무 늦게 pr을 확인한 것 같아요. 담에 올리시면 더 빨리 확인하도록 할게요

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/dto/request/GoalUpdateRequest.java:
```diff
@@ -20,6 +21,7 @@
+@ValidGoalPeriod
 @ValidTags
 public class GoalUpdateRequest implements TagValidatable {
 
```
implements와 아래 get 메서드 추가 하셔야 될 것 같습니다.

rlqor9688 replied:
감사합니다! 반영해서 올리겠습니다!

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/RetirementReferenceValueService.java:
```diff
@@ -53,15 +54,20 @@ public void init() throws IOException {
-
 		Long expectedLifespan = getExpectedLifespan(user.getGender(), currentYear);
 		RetirementType retirementType = RetirementType.COUPLE;
-		Long MonthlyExpense = getAverageMonthlyExpense(retirementType);
```
default 은퇴 목표 생성이 될 때 COUPLE 기준으로 월 지출이 계산이 되는 건가요?

rlqor9688 replied:
네 맞습니다! 부부 기준 월지출을 default로 설정했습니다.

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/GoalServiceImpl.java:
```diff
@@ -75,12 +77,20 @@ public GoalInfoResponse findGoal(Long goalId, Long userId) {
-			.stream()
+
+		List<Goal> goals = goalRepository.searchGoalsWithCursor(user.getId(), condition);
+
```
어떻게 condition.size()를 하나 했더니 recode로 구성되어있으면 가능하군요 하나 배웠습니다.

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/GoalServiceImpl.java:
```diff
@@ -75,12 +77,20 @@ public GoalInfoResponse findGoal(Long goalId, Long userId) {
+		Long nextCursor = hasNext ? result.get(result.size() -1).getId() : null;
+
+		return new CursorPageResponse<>(result, nextCursor, hasNext);
 	}
```
record가 모든 필드를 가진 생성자도 만들어주는 군요👍

byeongtaek12 commented on src/test/java/org/example/lifechart/domain/goal/service/DefaultRetirementGoalServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+		assertThat(response).isNotNull();
+		assertThat(response.getGoalId()).isEqualTo(1L);
+
+		ArgumentCaptor<GoalCreateRequest> captor = ArgumentCaptor.forClass(GoalCreateRequest.class);
```
이 부분은 // given에 두는 것이 맞는 것 같습니다.
![image](https://github.com/user-attachments/assets/318269c5-cddd-46c1-b689-1f67fad584fc)

rlqor9688 replied:
피드백 감사합니다! 반영했습니다!

byeongtaek12 reviewed (approved):
확인했습니다. 확실히 코드 리뷰를 하면 할 수록 뭔가 쌓이는 게 느껴지는 것 같습니다. 코멘트들만 한번 확인해주시면 될 것 같습니다.

rlqor9688 committed:
[fix] PR 피드백 반영
[Service&Test]
1. DefaultRetirementGoalService : 인스턴스를 정의하지 않고, builder에 반환값에서 직접 값을 받아 입력해주는 방식으로 리팩토링
2. DefaultRetirementGoalServiceTest : ArgumentCaptor는 테스트를 위한 준비 도구이기에 given에 두는 것이 적절하다는 피드백 반영
3. OpenApiApartmentPriceService : 주석 오타 수정(60개월 -> 120개월)
4. OpenApiApartmentPriceServiceTest : 예외 상황 테스트코드 추가
[Dto]
1. GoalUpdateRequest : HasGoalPeriod 상속, 메서드 override

rlqor9688 committed:
[fix] 피드백 반영 커밋
피드백 반영

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-goal

rlqor9688 committed:
[fix] 피드백 반영

rlqor9688 committed:
Merge branch 'feature/kiback-goal' of https://github.com/LifeChart19/LifeChart into feature/kiback-goal

rlqor9688 committed:
[fix] build.gradle 의존성 중복 제거
